docs/DevGuide/TagsTypelists.md

```diff
@@ -0,0 +1,190 @@
+\cond NEVER
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+# Tags and Type Lists {#tags_and_type_lists}
+
+\tableofcontents
+
+# Introduction to Tags {#documentation_for_tags}
+In a small C++ program, it is common to use the built-in fundamental types
+(bool, int, double, etc.) in computations, and to give instances of types
+variable names which refer to them. For example, a section of a small
+program may look like this:
+
+~~~{.c}
+const double pressure = 1.2;
+const double density = 0.5;
+
+std::map<std::string, double> naive_databox;
+naive_databox["Pressure"] = pressure;
+naive_databox["Density"] = density;
+
+return naive_databox["MisspelledKey"];
+~~~
+
+As you may have noticed, the computer is asked to return the value given a key
+that does not exist! As written however, the program is well-formed and there
+is no error emitted at compilation. Because the keys are indistinguishable
+from their type alone, the mistake is only caught at runtime. The runtime
+portion of a SpECTRE calculation will typically be much longer (up to thousands
+of times longer!) than the compile time portion, so it is critical to catch
+costly mistakes like this as early into the calculation as possible.
+SpECTRE takes advantage of C++'s static typing of to help users distinguish
+between quantities at the type level. This is done through tags.
+
+An example of a tag looks like the following:
+
+~~~{.c}
+struct Pressure {
+  //The type of the quantity to which the tag refers:
+  using type = double;
+};
+
+~~~
+
+Note that through the use of tags, we have now promoted `Pressure` from being
+a *value* used to initialize a `std::string` at runtime, to a *type*
+distinguishable from other types at compile time. A large portion of SpECTRE is
+designed with the philosophy of enlisting the help of the compiler in assuring
+the correctness of our programs.
+
+# Tags in SpECTRE {#documentation_for_tags_in_spectre}
+
+## OptionTags {#documentation_for_option_tags}
+You'll find an example of option tags used in the SingletonHelloWorld example
+executable. For convenience, we provide it here:
+
+\snippet SingletonHelloWorld.hpp executable_example_options
+
+Option tags reflect options that are specifiable at runtime through YAML input
+files. Option tags are used to specify runtime settings for the Domain,
+AnalyticSolution, and EventsAndTriggers, among other things.
+
+## InboxTags {#documentation_for_inbox_tags}
+SpECTRE is designed to run on multiple processors in parallel. By default then,
+SpECTRE assumes that data created on one processor is not needed to be
+communicated to any other processor. However, some programs need data to be
+communicated between processors, such as flux data. An inbox tag follows the
+general pattern:
+
+~~~{.c}
+struct TypeRecieveTag {
+  using temporal_id = int;
+  using type = std::unordered_map<temporal_id, TypeToReceive>;
```

Comment:
I think more explanation would be helpful to understand what is expected in `temporal_id` and `TypeToReceive`.